### DISCUSSION TITLE
How to insert Knex when creating a custom type in NestJS

### DISCUSSION THREAD

**Discussion by bryan-vh** [mikro-orm/mikro-orm#1992] [Q&A] [answered]
I'm creating a custom type but I need to return knex.raw as suggested in issue #372, and I'm a bit confused on how to do that within the NestJS framework, since MikroORM is injected throughout the project.

Essentially, what is the best way to get knex within a custom type when using NestJS?

Any help would be greatly appreciated!

**Comment by B4nan** [comment, marked as answer]
You can't get the configured instance, but you don't need it just for creating raw query fragmnet, doing `knex({ client: 'pg' })` is enough for that. What is the exact use case? Because nowadays it should be prefered to go with `convertToJSValueSQL` and `convertToDatabaseValueSQL` rather than using `knex.raw()`. 

https://mikro-orm.io/docs/custom-types/#advanced-example---pointtype-and-wkt

**Reply by bryan-vh** [reply]
Well, I more or less am creating the GeometryType for making a PostGIS object in PostgreSQL, and I am trying to write a string that is 'ST_SetSRID(ST_MakePoint(my_long_here, my_lat_here))' but when compiled to SQL outputs that as a string rather than raw SQL syntax. So I'm a bit unsure of how to leave it as raw SQL if that makes sense.

EDIT: thank you for responding so quickly on this discussion and others, it really helps me get unblocked from making progress on my app

**Reply by B4nan** [reply]
The link is exactly about that use case, you should use `convertToDatabaseValueSQL` instead.

**Reply by bryan-vh** [reply]
Okay, so if I just return the value presented there it will transform it from that string into raw SQL?

**Reply by B4nan** [reply]
the output of both `convertToDatabaseValueSQL` and `convertToJSValueSQL` is raw sql

**Reply by bryan-vh** [reply]
Okay, so this is the error I'm getting after updating what I'm doing based on your definition: 

```
insert into "party" ("address", "description", "end", "geom", "image", "latitude", "longitude", "name", "privacy", "start", "user_id", "uuid") values (NULL, NULL, '2008-10-31T17:04:32.000Z', 'ST_SetSRID(ST_MakePoint(120.37, 47.56))', 'https://google.com', '47.56', '120.37', 'bryan_0', 2, '2008-10-31T17:04:32.000Z', 'bryan', 'cafd6f56-2cb6-41c0-a084-93bff84190ac') returning "uuid" - parse error - invalid geometry
```

and this is the code I have:

```
export class Geometry {
  constructor(readonly latitude: number, readonly longitude: number) {}
}

export class GeometryType extends Type<Geometry | undefined, string> {
  convertToDatabaseValue(value: Geometry | undefined): string {
    if (!value) {
      throw ValidationError.invalidType(Geometry, value, 'js')
    }

    return `ST_SetSRID(ST_MakePoint(${value.longitude}, ${value.latitude}))`
  }

  convertToDatabaseValueSQL(key: string): string {
    return key
  }

  getColumnType(): string {
    return 'geometry(Point, 4326)'
  }
}
```

**Reply by B4nan** [reply]
nope, that's not how it works, check the link i sent you and do it based on that

**Reply by bryan-vh** [reply]
yeah I just realized my mistake, I was doing the ST commands within the first function rather than the second, thank you!
